### PR TITLE
Test data generator (#282)

### LINKED ISSUE #282: Scale testing from DB-to-front-end
**Problem**
Test the performance, at scale, of the Database and front-end transactions (query performance).

We need to test the mirror node for 1 year of anticipated data (in the database). At 4 TPS (mainnet throughput), we will have ~ 125M transactions, and ~ 625M cryptotransferlist entries. We should create tests that validate the infrastructure including DB and queries to ensure we can handle this volume and the rest-api queries still perform well against the DB with that much data in it.

### PR DESCRIPTION
**Pull request by apeksharma**
Signed-off-by: Apekshit Sharma <apekshit.sharma@hedera.com>

**Detailed description**:
See /faker/docs/README.md

**Which issue(s) this PR fixes**:
Fixes #282 

**Special notes for your reviewer**:
Will work on tests on tomorrow. Putting the review out earlier.

**Checklist**
- [x] Documentation added
- [ ] Tests updated

----
Sample output
```
appy apekshits-MacBook-Pro ~/git/hedera-mirror-node (faker2) → ./faker/run.sh
********************************************
Generating fake data files in dir /var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde
********************************************

           ,:if0@@@@@@@0f1;,.
       :L@@@@@@@@@@@@@@@@@@@@@C;
     ,L@@@@@@@@@@@@@@@@@@@@@@@@@C:
    ;0@@@@@@@@@@@@@@@@@@@@@@@@@@@81
   :0@@@@@@@1 10@@@@@@@@1 1@@@@@@@8i             1@L        ;@L.                      @@t
  ,L@@@@@@@@1 1@@@@@@@@@1 1@@@@@@@@0:            1@L        ;@L.                      @@t
  ;@@@@@@@@@1  _________  1@@@@@@@@@1            1@L        ;@L.   i@@@@@;     :@@@@@;@@t   .t@@@@f,   i@@@@@@t. .f@@@@@@@i  TM
 .1@@@@@@@@@1 1@@@@@@@@@1 1@@@@@@@@@C.           1@0000000000@L. :@@;  :f@G,  C@@:'':0@@t  1@@:  :@@f. t@@@::::' t8@1'':L@@t
 .1@@@@@@@@@1 1@@@@@@@@@1 1@@@@@@@@@C.           1@@@@@@@@@@@@L. C@@@@@@@@@t i@@.     @@t ;8@@@@@@@@81 t@C      ;0@.     L@t
  ;@@@@@@@@@1  _________  1@@@@@@@@@1            1@L        ;@L. C@:         i@@.     @@t ;@@:         t@C      ;0@.    .L@t
  ,L@@@@@@@@1 1@@@@@@@@@1 1@@@@@@@@0:            1@L        ;@L. ;@@@i:;t@@;  L@@1::i@@@t  1@@t;,;f8C: t@C       t@@t;:iC@@t
   :G@@@@@@@1 1@@@@@@@@@1 1@@@@@@@8;             ;@1        :@i    ;@@@@@i.    :t@@@@;tL;   ,i@@@@t;.  i@1        ,1@@@L;1@i
    :G@@@@@@@@@@@@@@@@@@@@@@@@@@@8i
     .f8@@@@@@@@@@@@@@@@@@@@@@@@L,
        ,L8@@@@@@@@@@@@@@@@@@0C:                                          Mirror Node  (v0.4.0-SNAPSHOT)
           ':if0@@@@@@@0fi:'

2019-11-20 18:40:35,671 INFO  [main ] c.h.f.Faker Starting Faker v0.4.0-SNAPSHOT on apekshits-MacBook-Pro.local with PID 41229 (/Users/appy/git/hedera-mirror-node/target/mirror-node-0.4.0-SNAPSHOT.jar started by appy in /Users/appy/git/hedera-mirror-node)
2019-11-20 18:40:35,673 INFO  [main ] c.h.f.Faker No active profile set, falling back to default profiles: default
2019-11-20 18:40:37,401 INFO  [main ] c.h.f.d.w.PostgresCSVDomainWriter Writing CSV files to /var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde
2019-11-20 18:40:37,794 INFO  [main ] c.h.f.Faker Started Faker in 2.444 seconds (JVM running for 3.227)
2019-11-20 18:40:37,794 INFO  [main ] c.h.f.d.DomainDriver Simulation time from 1577836800 to 1577845440 (time period: 8640sec)
2019-11-20 18:40:37,889 INFO  [main ] c.h.f.d.DomainDriver Generated 10000 transactions in 91.71 ms
2019-11-20 18:40:37,963 INFO  [main ] c.h.f.d.DomainDriver Generated 20000 transactions in 165.7 ms
2019-11-20 18:40:38,028 INFO  [main ] c.h.f.d.DomainDriver Generated 30000 transactions in 231.4 ms
2019-11-20 18:40:38,094 INFO  [main ] c.h.f.d.DomainDriver Generated 40000 transactions in 297.2 ms
2019-11-20 18:40:38,117 INFO  [main ] c.h.f.d.DomainDriver Generated 43177 transactions in 320.0 ms
2019-11-20 18:40:38,317 INFO  [main ] c.h.f.d.g.e.EntityGenerator Wrote 10289 active and 27 deleted account entities
2019-11-20 18:40:38,321 INFO  [main ] c.h.f.d.g.e.EntityGenerator Wrote 413 active and 0 deleted file entities
2019-11-20 18:40:38,321 INFO  [main ] c.h.f.d.g.e.EntityGenerator Wrote 0 active and 0 deleted contract entities
2019-11-20 18:40:38,322 INFO  [main ] c.h.f.d.g.e.EntityGenerator Generated all entities in 103.3 ms
2019-11-20 18:40:38,323 INFO  [main ] c.h.f.d.DomainDriver Total time taken: 526.1 ms


***********************************************
Loading fake data files into PostgresSQL Server
***********************************************
------------------------------
Truncating tables
------------------------------
psql:/var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/load_data.sql:4: NOTICE:  truncate cascades to table "t_transactions"
psql:/var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/load_data.sql:4: NOTICE:  truncate cascades to table "t_cryptotransferlists"
psql:/var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/load_data.sql:4: NOTICE:  truncate cascades to table "t_contract_result"
psql:/var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/load_data.sql:4: NOTICE:  truncate cascades to table "t_livehashes"
TRUNCATE TABLE
psql:/var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/load_data.sql:5: NOTICE:  truncate cascades to table "t_transactions"
psql:/var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/load_data.sql:5: NOTICE:  truncate cascades to table "t_cryptotransferlists"
psql:/var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/load_data.sql:5: NOTICE:  truncate cascades to table "t_contract_result"
psql:/var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/load_data.sql:5: NOTICE:  truncate cascades to table "t_livehashes"
TRUNCATE TABLE
psql:/var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/load_data.sql:6: NOTICE:  truncate cascades to table "t_cryptotransferlists"
psql:/var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/load_data.sql:6: NOTICE:  truncate cascades to table "t_contract_result"
psql:/var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/load_data.sql:6: NOTICE:  truncate cascades to table "t_livehashes"
TRUNCATE TABLE
TRUNCATE TABLE
TRUNCATE TABLE
TRUNCATE TABLE
------------------------------
COPY data to t_record_files from /var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/t_record_files
------------------------------
COPY 1
------------------------------
COPY data to t_entities from /var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/t_entities
------------------------------
COPY 10729
------------------------------
COPY data to t_transactions from /var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/t_transactions
------------------------------
Dropping all constraints and indexes
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
DROP INDEX
DROP INDEX
DROP INDEX
DROP INDEX
Copying data
COPY 43177
Recreate all constraints and indexes
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
ALTER TABLE
------------------------------
COPY data to t_cryptotransferlists from /var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/t_cryptotransferlists
------------------------------
Dropping all constraints and indexes
ALTER TABLE
ALTER TABLE
DROP INDEX
DROP INDEX
DROP INDEX
Copying data
COPY 214736
Recreate all constraints and indexes
CREATE INDEX
CREATE INDEX
CREATE INDEX
ALTER TABLE
ALTER TABLE
------------------------------
COPY data to t_file_data from /var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/t_file_data
------------------------------
Dropping all constraints and indexes
DROP INDEX
Copying data
COPY 413
Recreate all constraints and indexes
CREATE INDEX
------------------------------
COPY data to account_balances from /var/folders/tt/5qkdxvqd2jsg5bdx2fskzbdc0000gn/T/tmp.G4E2UDde/account_balances
------------------------------
Dropping all constraints and indexes
ALTER TABLE
DROP INDEX
Copying data
COPY 86541
Recreate all constraints and indexes
CREATE INDEX
ALTER TABLE
CREATE INDEX
 count
-------
 10729
(1 row)

 count
-------
 43177
(1 row)

 count
--------
 214736
(1 row)

 count
-------
   413
(1 row)

 count
-------
 86541
(1 row)

 count
-------
     1
(1 row)
```

### CODE REVIEW

**Comment by apeksharma** [comment]
> Move the foreign key dropping/adding sql to a sproc in importer's DB. It's more likely to be updated there and is useful outside of just the faker.

Done.

> Replicating the logic of HAPI isn't needed, just the scale of the number of records added to the database.

I believe you meant the parts of active/deleted entities and RandomSet.
Simplified it.
Removed tracking of active entities. Still need to track deleted entities to have some deleted entities in t_entities in the end.
Removed RandomSet class.
However, am continuing to ensure things like not creating CRYPTOUPDATE transaction for a file entity (by bucketing ids for different entity types)


> If you want to show the impact of steadily increasing accounts/entities - generate the transactions in the DB, replicate it for a month, add some more accounts and do it again for another month.

Nah, don't want to do that right now.

Rest of the comments basically focus on two parts: *generation* and *loading* of fake data

#### Generation
> Rather than that frequency distribution code, just configure the relative frequencies of each transaction type (createAccount 10, updateAccount 1, cryptotransfer 1000...), generate all those transactions (1011 in this case), shuffle the list ....
> Guaranteed to have the proper distribution frequency that way.

Done. Replaced probabilistic frequency distribution with your suggestion above.

#### Loading
> There isn't a need to go the java to CSV to psql route if a small set of data is added directly from java to postgres and then that data is replicated (with different consensus timestamps) in the database.

You're right that csv is not needed for small set of data, and directly from java to postgres can be done. 
But for scalability testing, generating one year of data at @5tps and ~5crypto transfer per cypto transaction would be much more than a billion rows (including transactions, ctl, account balances).
COPY is the fastest way to load data. Reading blogs, people report like 3x faster speeds. That's because 
there is no network latency per batch; no parsing, planning, sql optimizing of million statements; less commits and fsyncs to disk, etc.
In my tests, loading 2 days worth of fake data took ~3min. That translates to ~6-8 hrs to load an year worth of fake data. If COPY is even 2x fast, that's a lot of saving in test execution time.

Another major benefit of CSV files is - golden data.
To be able to measure performance changes for say 2 different systems configurations, or across nightly changes, a golden set of data with baseline numbers is mandatory.

So when I had the choice to do one of the two - load directly from java to postgres, or csv + copy -  i choose later.
But i can see that pretty soon we'll have the direct version too as it has it's benefits too.

**Comment by apeksharma** [comment]
By no means this is the final version of this code. We'll be adding many tests in upcoming months: 
a) grpc api tests
b) importer scalability tests
c) nightly performance tests
etc
So it should be okay to get current code in, start using it, and make improvements in future iterations. Just don't want us to get blocked trying to make this new component which works, perfect. We have enough on our hands.

**Comment by apeksharma** [comment]
Nana and Steven would prefer a single PR with master as base, so switching to one PR (#398 ) rather than two (this and #398 ).

File: .circleci/config.yml [hunk 1]
```diff
@@ -73,7 +73,9 @@ jobs:
             NAME=hedera-mirror-importer-${CIRCLE_TAG:-b$CIRCLE_BUILD_NUM}
             WORKSPACE=/tmp/workspace
             mkdir -p ${WORKSPACE}/${NAME}
-            mv hedera-mirror-importer/target/hedera-mirror-importer-*.jar ${WORKSPACE}/${NAME}
+            # Truncate '-exec' from jar name
+            jarname=jarname=$(ls hedera-mirror-importer/target/hedera-mirror-importer-*-exec.jar | sed -E "s:.*target/(.*)-exec.jar:\1:")
+            mv hedera-mirror-importer/target/hedera-mirror-importer-*-exec.jar ${WORKSPACE}/${NAME}/${jarname}.jar
```
Comment (steven-sheehy) [inline .circleci/config.yml]: Can't you just `mv hedera-mirror-importer/target/hedera-mirror-importer-*-exec.jar ${WORKSPACE}/${NAME}/${NAME}.jar`?

File: hedera-mirror-faker/src/main/java/com/hedera/faker/FakerProperties.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+@Named
+@ConfigurationProperties("faker")
+public class FakerProperties {
+    private NumberDistributionConfig transactionsPerSecond = new NumberDistributionConfig();
+
+    /** Start time for the fake transactions. Used for consensus timestamp field. */
+    private int startTimeSec = 1577836800;  // 1 Jan 2020, 00:00:00
+
+    /** Fake transactions will be generated for time period from startTimeSec to startTimeSec + totalTimeSec. */
+    private int totalTimeSec = 3600 * 24;  // 1 hr
```
[the remark this answers is not in the document]
Comment (apeksharma) [reply hedera-mirror-faker/src/main/java/com/hedera/faker/FakerProperties.java]: Cool.

**Pull request closed without merging**